File: internal/httpapi/audit.go
package httpapi

import (
	"strconv"
	"time"

	"github.com/gin-gonic/gin"

	"github.com/myGithub/mcp-proxy-gateway/internal/audit"
	"github.com/myGithub/mcp-proxy-gateway/internal/domain"
	"github.com/myGithub/mcp-proxy-gateway/internal/store"
)

// 本文件实现审计日志分页查询端点（Req 22.4、17.5）：
//
//   GET /api/admin/audit?page=&pageSize=   按发生时间倒序分页返回审计记录及总数
//
// 分页参数通过查询参数 page/pageSize 传入，均为可选整数：
//   - page 缺省或非正由审计服务归正为第 1 页；
//   - pageSize 缺省或非正由审计服务取配置默认值（audit.page_size_default），
//     越界值由其收敛到 [1,200]（Req 22.4）。
// 本层仅负责整数解析（非整数返回字段级 VALIDATION），倒序与收敛由审计服务保证。

// registerAuditRoutes 在管理分组下注册审计日志查询端点（Req 22.4）。
func (r *Router) registerAuditRoutes(g *gin.RouterGroup) {
	g.GET("/audit", r.queryAudit)
	g.GET("/audit/export", r.exportAudit)
}

// queryAudit 按发生时间倒序分页返回审计记录（Req 22.4）。
func (r *Router) queryAudit(c *gin.Context) {
	if r.audit == nil {
		respondServiceUnavailable(c, "审计查询服务未就绪")
		return
	}
	page, ok := parseOptionalInt(c, "page")
	if !ok {
		return
	}
	pageSize, ok := parseOptionalInt(c, "pageSize")
	if !ok {
		return
	}
	query, ok := parseAuditQuery(c)
	if !ok {
		return
	}
	result, err := r.audit.List(c.Request.Context(), page, pageSize, query)
	if err != nil {
		respondError(c, err)
		return
	}
	respondOK(c, gin.H{
		"records":  result.Records,
		"page":     result.Page,
		"pageSize": result.PageSize,
		"total":    result.Total,
	})
}

func (r *Router) exportAudit(c *gin.Context) {
	if r.audit == nil {
		respondServiceUnavailable(c, "审计查询服务未就绪")
		return
	}
	page, ok := parseOptionalInt(c, "page")
	if !ok {
		return
	}
	pageSize, ok := parseOptionalInt(c, "pageSize")
	if !ok {
		return
	}
	query, ok := parseAuditQuery(c)
	if !ok {
		return
	}
	result, err := r.audit.List(c.Request.Context(), page, pageSize, query)
	if err != nil {
		respondError(c, err)
		return
	}
	respondJSONDownload(c, "mpg-audit-"+time.Now().Format("20060102-150405")+".json", result.Records)
}

func parseAuditQuery(c *gin.Context) (audit.Query, bool) {
	var query audit.Query
	query.EventType = c.Query("eventType")
	if !validAuditEventType(query.EventType) {
		respondError(c, domain.NewValidationError("审计事件类型参数非法", map[string]string{
			"eventType": "仅支持 login、create、update、delete、access_denied",
		}))
		return audit.Query{}, false
	}

	start, ok := parseOptionalRFC3339(c, "start")
	if !ok {
		return audit.Query{}, false
	}
	end, ok := parseOptionalRFC3339(c, "end")
	if !ok {
		return audit.Query{}, false
	}
	if !start.IsZero() && !end.IsZero() && start.After(end) {
		respondError(c, domain.NewValidationError("审计时间范围参数非法", map[string]string{
			"start": "开始时间不得晚于结束时间",
		}))
		return audit.Query{}, false
	}
	query.Start = start
	query.End = end
	return query, true
}

func validAuditEventType(eventType string) bool {
	switch eventType {
	case "", store.AuditEventLogin, store.AuditEventCreate, store.AuditEventUpdate, store.AuditEventDelete, store.AuditEventAccessDenied:
		return true
	default:
		return false
	}
}

func parseOptionalRFC3339(c *gin.Context, name string) (time.Time, bool) {
	value := c.Query(name)
	if value == "" {
		return time.Time{}, true
	}
	t, err := time.Parse(timeLayout, value)
	if err != nil {
		respondError(c, domain.NewValidationError("时间参数非法", map[string]string{
			name: "时间格式需为 RFC3339",
		}))
		return time.Time{}, false
	}
	return t, true
}

// parseOptionalInt 解析可选整数查询参数；缺省返回 0（交由下层收敛），非整数则以
// 字段级 VALIDATION 响应并返回 ok=false（调用方据此中止）。
func parseOptionalInt(c *gin.Context, name string) (int, bool) {
	v := c.Query(name)
	if v == "" {
		return 0, true
	}
	n, err := strconv.Atoi(v)
	if err != nil {
		respondError(c, domain.NewValidationError("分页参数非法", map[string]string{
			name: "需为整数",
		}))
		return 0, false
	}
	return n, true
}

File: web/src/api/audit.ts
/**
 * 审计日志查询 API 封装（任务 26.5）
 *
 * 设计要点（对应 internal/httpapi/audit.go 与 Req 22.4、17.5）：
 * - 端点挂载于管理前缀 `/api/admin/audit`（见 internal/httpapi/audit.go）；
 * - 复用全局 Axios 实例（`@/api/request`，baseURL=`/api/admin`），自动注入 JWT 并处理 401（Req 17.6）；
 * - 按发生时间倒序分页：page 从 1 起、pageSize 默认 20、范围 1-200，越界由后端收敛（Req 22.4）。
 *
 * 后端真实路由（已实现，见 internal/httpapi/audit.go）：
 *   GET /audit?page=&pageSize=   按 occurred_at 倒序分页返回审计记录及总数
 *
 * 响应字段名与后端 Go 结构体一致：记录字段为首字母大写（store.AuditRecord 无 json tag），
 * 分页元信息为小写（gin.H 中显式指定 page/pageSize/total）。
 */
import request from '@/api/request'

/** 审计事件类型，与后端 AuditEvent* 常量对齐。 */
export type AuditEventType = 'login' | 'create' | 'update' | 'delete' | 'access_denied'

/** 审计事件类型的中文显示名映射。 */
export const AUDIT_EVENT_LABELS: Record<string, string> = {
  login: '登录',
  create: '创建',
  update: '更新',
  delete: '删除',
  access_denied: '访问被拒',
}

/**
 * 一条审计日志记录（与后端 store.AuditRecord 对齐，无 json tag 故首字母大写）。
 */
export interface AuditRecord {
  /** 审计记录自增标识。 */
  ID: number
  /** 事件类型（见 AuditEventType）。 */
  EventType: string
  /** 操作目标对象（如上游名称、规则标识）；可为空。 */
  Target: string
  /** 事件明细的结构化 JSON（如登录结果、变更字段）；可为空（null）。 */
  Detail: unknown
  /** 事件发生时间（RFC3339）。 */
  OccurredAt: string
}

/** 审计分页查询结果（与后端 queryAudit 响应体对齐）。 */
export interface AuditPage {
  /** 本页记录，按发生时间倒序；无数据时为空数组。 */
  records: AuditRecord[]
  /** 生效页码（从 1 起，已对非法入参收敛）。 */
  page: number
  /** 生效的每页条数（已收敛到 1-200）。 */
  pageSize: number
  /** 审计记录总数，供调用方计算总页数。 */
  total: number
}

export interface AuditQuery {
  page?: number
  pageSize?: number
  eventType?: string
  start?: string
  end?: string
}

/** GET /audit 的原始响应体；records 可能为 null，归一化为空数组。 */
interface AuditResponse {
  records: AuditRecord[] | null
  page: number
  pageSize: number
  total: number
}

/** 审计分页每页条数下界（与后端一致）。 */
export const AUDIT_MIN_PAGE_SIZE = 1
/** 审计分页每页条数上界（与后端一致）。 */
export const AUDIT_MAX_PAGE_SIZE = 200
/** 审计分页每页条数默认值（与后端一致）。 */
export const AUDIT_DEFAULT_PAGE_SIZE = 20

/**
 * 按发生时间倒序分页查询审计记录（Req 22.4）。
 *
 * page/pageSize 均为可选：缺省或非正由后端收敛（page→1，pageSize→默认 20），
 * pageSize 超过 200 收敛为 200。响应回显实际生效的 page/pageSize 与总数。
 */
export async function listAudit(query: AuditQuery = {}): Promise<AuditPage> {
  const params: Record<string, string> = {}
  if (query.page !== undefined && query.page > 0) {
    params.page = String(query.page)
  }
  if (query.pageSize !== undefined && query.pageSize > 0) {
    params.pageSize = String(query.pageSize)
  }
  if (query.eventType !== undefined && query.eventType !== '') {
    params.eventType = query.eventType
  }
  if (query.start !== undefined && query.start !== '') {
    params.start = query.start
  }
  if (query.end !== undefined && query.end !== '') {
    params.end = query.end
  }
  const res = await request.get<AuditResponse>('/audit', { params })
  return {
    records: res.data?.records ?? [],
    page: res.data?.page ?? 1,
    pageSize: res.data?.pageSize ?? AUDIT_DEFAULT_PAGE_SIZE,
    total: res.data?.total ?? 0,
  }
}
